src/V3Hasher.cpp

```diff
@@ -29,40 +31,290 @@ class HasherVisitor final : public AstNVisitor {
-                // For identical nodes, the type should be the same thus
-                // dtypep should be the same too
-                m_lowerHash = V3Hash(m_lowerHash,
-                                     V3Hash(V3Hash(nodep->type() << 6), V3Hash(nodep->dtypep())));
-                // Now update m_lowerHash for our children's (and next children) contributions
-                iterateChildrenConst(nodep);
-                // Store the hash value
-                if (m_cacheInUser4) { nodep->user4(m_lowerHash.fullValue()); }
-                thisHash = m_lowerHash;
-            }
+    inline void walk(AstNode* nodep) {  //
+        if (VL_LIKELY(nodep)) iterate(nodep);
+    }
+
+    V3Hash hashNode(AstNode* nodep, bool hashDType, bool hashChildren, std::function<void()>&& f) {
+        if (m_cacheInUser4 && nodep->user4()) {
+            return V3Hash(nodep->user4());
+        } else {
+            VL_RESTORER(m_hash);
+            // Reset accumulator
+            m_hash = V3Hash(nodep->type());  // Node type
+            f();  // Node specific hash
+            if (hashDType && nodep != nodep->dtypep()) walk(nodep->dtypep());  // Node dtype
+            if (hashChildren) iterateChildrenConst(nodep);  // Children
+            if (m_cacheInUser4) { nodep->user4(m_hash.value()); }
+            return m_hash;
         }
-        // Update what will become the above node's hash
-        m_lowerHash += m_cacheInUser4 ? V3Hash(nodep->user4()) : thisHash;
     }
 
+    // VISITORS
+
+    // Each visitor below contributes to the hash any node specific content
+    // that is not dependent on either of the following, as these are
+    // included by default by hashNode:
+    // - Node type (as given by AstNode::type())
+    // - Node dtype (unless !hashDType)
+    // - child nodes (unless !hashChildren)
+    //
+    // The hash must be stable, which means in particular it cannot rely on
+    // pointer values, or any other value that might differ between separate
+    // invocations of Verilator over the same design.
+    //
+    // Note there is a circularity problem where some child nodes can back
+    // to their ancestral nodes via member pointers, which can lead to an
+    // infinite traversal. To break this, nodes that are subject to such
+    // referencing and represent code which can reasonably be assumed not to
+    // be equivalent to any other code, are hashed either by name (e.g.:
+    // AstNodeModule), or by unique identifier (e.g.: AstNodeUOrStructDType).
+
+#define VISIT(kind, hashDType, hashChildren, body) \
+    virtual void visit(kind* nodep) override { \
+        const V3Hash hash = hashNode(nodep, hashDType, hashChildren, [=]() body); \
+        m_hash += hash; \
+    }
+
+    //------------------------------------------------------------
+    // AstNode - Warns to help find missing cases
+
+    VISIT(AstNode, true, true, {  //
+        UINFO(4, "WARNING! Hashing node as AstNode: " << nodep);
+    })
```

Comment:
Adding macros like this seems a bit hard to read, I think I'd prefer keeping the normal void visit first line, e.g.

```suggestion
    virtual void visit(AstNode*) {
        m_hash += hashNodeAndIterate(true, true);
        UINFO(4, "WARNING! Hashing node as AstNode: " << nodep);
    }
```

This is an extra line but more obvious to casual reading, and makes it more like other visitor classes.
BTW try to put "AndIterate" in those functions that iterate as avoids bugs where forgetting that iterate has already been called in a visitor.

Reply:
That's not really the expansion of this macro. I expanded them, see how you like them.